### DISCUSSION TITLE
[item01] valueOf 는 어떤 경우에 사용될까요?

### DISCUSSION THREAD

**Discussion by cueon** [woowacourse-7th-study/effective-java#3] [질문] [answered]
from, of 의 차이점은 바로 보이는데
이 둘과 valueOf 의 차이점은 확실하게 구분되지 않는다는 느낌을 받았습니다

이 부분에 대해 이야기를 나눠보고 싶습니다

from, of 와 valueOf 의 차이점을 어떻게 이해하셨나요?

**Comment by JHZLO** [comment, marked as answer]
저도 공부하면서 단순히 valueOf가 from과 of의 자세한 버전이라길래 많이 헷갈렸던 것 같습니다..

그래서 조금 찾아본 결과 다음과 같이 이해하였습니다.

```java
public enum Color {
  RED,
  BLUE;
}
...
Color redColor = Color.valueOf("RED");
Color blueColor = Color.valueOf("BLUE");
```
위와 같이 `valueOf`의 경우에는 enum의 요소를 "조회"할 때에 사용되는 메서드 입니다.

미리 생성된 객체를 "조회"하는 메서드이기 때문에 팩토리의 역할을 한다고 볼 수는 없지만, 

외부에서 원하는 객체를 반환해주고 있으므로 결과적으로는 정적 팩토리 메서드라고 간주할 수 있습니다.

즉, from과 of와 달리 `valueOf`는 enum에서만 쓰이는 특정한 정팩메로 이해하면 좋을 것 같습니다!

**Reply by cueon** [reply]
준형님이 말씀해주신 내용으로 이해해도 좋겠네요!

```java
Integer.valueOf("123");
```

이런 경우도 enum 을 사용했다고 볼 수 있을까요?

```java
Boolean.valueOf("TrUe");
```

이런 경우도 허용된다고 합니다

너무 모호한 부분 같습니다..

**Reply by JHZLO** [reply]
아, 그럼 조금 더 범용적인 관점으로 접근해서,
valueOf는 Enum, Wrapper 클래스에서 객체를 반환할 때 사용되는 정적 팩토리 메서드로 이해해야 하는 게 좋을까요?
